File: src/main/java/io/zeitwert/fm/contact/model/base/ObjContactPartAddressBase.java
package io.zeitwert.fm.contact.model.base;

import io.zeitwert.fm.contact.model.ObjContact;
import io.zeitwert.fm.contact.model.ObjContactPartAddress;
import io.zeitwert.fm.contact.model.enums.CodeAddressChannel;
import io.zeitwert.ddd.obj.model.base.ObjPartBase;
import io.zeitwert.ddd.oe.model.enums.CodeCountry;
import io.zeitwert.ddd.part.model.PartRepository;
import io.zeitwert.ddd.property.model.EnumProperty;
import io.zeitwert.ddd.property.model.SimpleProperty;

public abstract class ObjContactPartAddressBase extends ObjPartBase<ObjContact> implements ObjContactPartAddress {

	protected final EnumProperty<CodeAddressChannel> addressChannel = this.addEnumProperty("addressChannel",
			CodeAddressChannel.class);
	protected final SimpleProperty<String> name = this.addSimpleProperty("name", String.class);
	protected final SimpleProperty<String> street = this.addSimpleProperty("street", String.class);
	protected final SimpleProperty<String> zip = this.addSimpleProperty("zip", String.class);
	protected final SimpleProperty<String> city = this.addSimpleProperty("city", String.class);
	protected final EnumProperty<CodeCountry> country = this.addEnumProperty("country", CodeCountry.class);

	public ObjContactPartAddressBase(PartRepository<ObjContact, ?> repository, ObjContact obj, Object state) {
		super(repository, obj, state);
	}

}

File: fm-domain/src/main/java/io/zeitwert/fm/contact/model/base/ObjContactPartAddressBase.kt
package io.zeitwert.fm.contact.model.base

import io.dddrive.ddd.model.PartRepository
import io.dddrive.obj.model.base.ObjPartBase
import io.dddrive.property.model.Property
import io.zeitwert.fm.contact.model.ObjContact
import io.zeitwert.fm.contact.model.ObjContactPartAddress
import io.zeitwert.fm.contact.model.enums.CodeAddressChannel
import io.zeitwert.fm.oe.model.enums.CodeCountry

/**
 * Base class for ObjContactPartAddress using the NEW dddrive framework.
 */
abstract class ObjContactPartAddressBase(
	obj: ObjContact,
	repository: PartRepository<ObjContact, ObjContactPartAddress>,
	property: Property<*>,
	id: Int,
) : ObjPartBase<ObjContact>(obj, repository, property, id),
	ObjContactPartAddress {

	override fun doInit() {
		super.doInit()
		addEnumProperty("addressChannel", CodeAddressChannel::class.java)
		addBaseProperty("name", String::class.java)
		addBaseProperty("street", String::class.java)
		addBaseProperty("zip", String::class.java)
		addBaseProperty("city", String::class.java)
		addEnumProperty("country", CodeCountry::class.java)
	}

	override val addressType get() = addressChannel?.addressType

	override val isMailAddress get() = addressChannel?.isMailAddress

}
